tests/journal_lock.py/test_journal_lock.py

```diff
@@ -0,0 +1,194 @@
+            import ntsecuritycon as con
+            import win32security
+
+            # Fetch user details
+            winuser, domain, type = win32security.LookupAccountName("", os.environ.get('USERNAME'))
+            # Fetch the current security of tmpdir for that user.
+            sd = win32security.GetFileSecurity(str(tmpdir), win32security.DACL_SECURITY_INFORMATION)
+            dacl = sd.GetSecurityDescriptorDacl()  # instead of dacl = win32security.ACL()
+
+            # Add Write to Denied list
+            # con.FILE_WRITE_DATA results in a 'Special permissions' being
+            # listed on Properties > Security for the user in the 'Deny' column.
+            # Clicking through to 'Advanced' shows a 'Deny' for
+            # 'Create files / write data'.
+            dacl.AddAccessDeniedAce(win32security.ACL_REVISION, con.FILE_WRITE_DATA, winuser)
+            # Apply that change.
+            sd.SetSecurityDescriptorDacl(1, dacl, 0)  # may not be necessary
+            win32security.SetFileSecurity(str(tmpdir), win32security.DACL_SECURITY_INFORMATION, sd)
+
+        else:
+            import stat
+            os.chmod(tmpdir, stat.S_IRUSR | stat.S_IXUSR)
+
+        jlock = JournalLock()
+
+        # Check that an actual journaldir is handled correctly.
+        locked = jlock.obtain_lock()
+
+        # Revert permissions for test cleanup
+        if sys.platform == 'win32':
+            # We can reuse winuser etc from before
+            import pywintypes
+
+            # We have to call GetAce() until we find one that looks like what
+            # we added.
+            i = 0
+            ace = dacl.GetAce(i)
+            while ace:
+                if ace[0] == (con.ACCESS_DENIED_ACE_TYPE, 0) and ace[1] == con.FILE_WRITE_DATA:
+                    # Delete the Ace that we added
+                    dacl.DeleteAce(i)
+                    # Apply that change.
+                    sd.SetSecurityDescriptorDacl(1, dacl, 0)  # may not be necessary
+                    win32security.SetFileSecurity(str(tmpdir), win32security.DACL_SECURITY_INFORMATION, sd)
+                    break
+
+                i += 1
+                try:
+                    ace = dacl.GetAce(i)
+
+                except pywintypes.error:
+                    print("Couldn't find the Ace we added, so can't remove")
+                    break
+
+        else:
+            os.chmod(tmpdir, stat.S_IRUSR | stat.S_IWUSR | stat.S_IXUSR)
+
+        assert locked == JournalLockResult.JOURNALDIR_READONLY
+
+    def test_obtain_lock_already_locked(self, mock_journaldir: py_path_local_LocalPath):
+        """Test JournalLock.obtain_lock() with tmpdir."""
+        jlock = JournalLock()
+
+        locked = jlock.obtain_lock()
+        assert locked == JournalLockResult.LOCKED
+        # Now attempt to lock again, but only that.
+        second_attempt = jlock._obtain_lock()
+        assert second_attempt == JournalLockResult.ALREADY_LOCKED
```

Comment:
Fails on Linux because flock(2) (which is what fcntl.flock() uses) is a *per process* lock.  As the test is still in the same process, it can lock again because it has the lock already.

Will need to look into multiprocessing to properly test this.